<a href="https://colab.research.google.com/github/shreehere/Machine-Learning--Task-7/blob/main/Task_7_Assignment_10_Euclidean%20and%20Manhattan%20Distance%20using%20KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##K- Nearest Neighbour (KNN) Classifier

In [1]:
# importing all the necessary Libraries

import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from scipy.spatial import distance

Reading the train dataset in Pandas DataFrame using read_csv

In [3]:
train_df=pd.read_csv("/content/trainKNN.txt",header=None)
train_df

0        1      2     3     4      5     6     7     8    9   10
0      1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0   1
1      2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0   1
2      3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0   1
3      4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0   1
4      5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0   1
..   ...      ...    ...   ...   ...    ...   ...   ...   ...  ...  ..
191  207  1.51645  14.94  0.00  1.87  73.11  0.00  8.67  1.38  0.0   7
192  208  1.51831  14.39  0.00  1.82  72.86  1.41  6.47  2.88  0.0   7
193  209  1.51640  14.37  0.00  2.74  72.85  0.00  9.45  0.54  0.0   7
194  210  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0   7
195  211  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0   7

[196 rows x 11 columns]

Changing the column names to the original data column names

In [5]:
train_df.columns=['Id','RI','Na','Mg','Al','Si','K','Ca','Ba','Fe','Type of glass']
train_df

Id       RI     Na    Mg    Al     Si     K    Ca    Ba   Fe  \
0      1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0   
1      2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0   
2      3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0   
3      4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0   
4      5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0   
..   ...      ...    ...   ...   ...    ...   ...   ...   ...  ...   
191  207  1.51645  14.94  0.00  1.87  73.11  0.00  8.67  1.38  0.0   
192  208  1.51831  14.39  0.00  1.82  72.86  1.41  6.47  2.88  0.0   
193  209  1.51640  14.37  0.00  2.74  72.85  0.00  9.45  0.54  0.0   
194  210  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0   
195  211  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0   

     Type of glass  
0                1  
1                1  
2                1  
3                1  
4                1  
..             ...  
191              7  
192              7  
193              7  
194              7  
195              7  

[196 rows x 11 columns]

Droping the 'Id' column since it is not required

In [7]:
train_df=train_df.drop('Id',axis=1)
train_df

RI     Na    Mg    Al     Si     K    Ca    Ba   Fe  Type of glass
0    1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0              1
1    1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0              1
2    1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0              1
3    1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0              1
4    1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0              1
..       ...    ...   ...   ...    ...   ...   ...   ...  ...            ...
191  1.51645  14.94  0.00  1.87  73.11  0.00  8.67  1.38  0.0              7
192  1.51831  14.39  0.00  1.82  72.86  1.41  6.47  2.88  0.0              7
193  1.51640  14.37  0.00  2.74  72.85  0.00  9.45  0.54  0.0              7
194  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0              7
195  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0              7

[196 rows x 10 columns]

Doing the same for the test data

In [9]:
test_df=pd.read_csv("/content/testKNN.txt",header=None)
test_df.columns=['Id','RI','Na','Mg','Al','Si','K','Ca','Ba','Fe','Type of glass']
test_df=test_df.drop('Id',axis=1)
test_df

RI     Na    Mg    Al     Si     K     Ca    Ba    Fe  Type of glass
0   1.52152  13.05  3.65  0.87  72.32  0.19   9.85  0.00  0.17              1
1   1.52152  13.12  3.58  0.90  72.20  0.23   9.82  0.00  0.16              1
2   1.52300  13.31  3.58  0.82  71.99  0.12  10.17  0.00  0.03              1
3   1.51709  13.00  3.47  1.79  72.72  0.66   8.18  0.00  0.00              2
4   1.51660  12.99  3.18  1.23  72.97  0.58   8.81  0.00  0.24              2
5   1.51839  12.85  3.67  1.24  72.57  0.62   8.68  0.00  0.35              2
6   1.51832  13.33  3.34  1.54  72.14  0.56   8.99  0.00  0.00              3
7   1.51934  13.64  3.54  0.75  72.65  0.16   8.89  0.15  0.24              3
8   1.52211  14.19  3.78  0.91  71.36  0.23   9.14  0.00  0.37              3
9   1.52043  13.38  0.00  1.40  72.25  0.33  12.50  0.00  0.00              5
10  1.52058  12.85  1.61  2.17  72.18  0.76   9.70  0.24  0.51              5
11  1.52119  12.97  0.33  1.51  73.39  0.13  11.27  0.00  0.28              5
12  1.51916  14.15  0.00  2.09  72.74  0.00  10.88  0.00  0.00              6
13  1.51969  14.56  0.00  0.56  73.48  0.00  11.22  0.00  0.00              6
14  1.51115  17.38  0.00  0.34  75.41  0.00   6.65  0.00  0.00              6
15  1.52065  14.36  0.00  2.02  73.42  0.00   8.44  1.64  0.00              7
16  1.51651  14.38  0.00  1.94  73.61  0.00   8.48  1.57  0.00              7
17  1.51711  14.23  0.00  2.08  73.36  0.00   8.62  1.67  0.00              7

In [10]:
train_df.describe()

RI          Na          Mg          Al          Si           K  \
count  196.000000  196.000000  196.000000  196.000000  196.000000  196.000000   
mean     1.518295   13.375204    2.758980    1.454337   72.635408    0.519388   
std      0.003055    0.783145    1.392641    0.491688    0.763578    0.672703   
min      1.511310   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516458   12.877500    2.362500    1.190000   72.317500    0.140000   
50%      1.517630   13.280000    3.480000    1.360000   72.810000    0.560000   
75%      1.518985   13.792500    3.610000    1.622500   73.080000    0.610000   
max      1.533930   15.790000    4.490000    3.500000   75.180000    6.210000   

               Ca          Ba          Fe  Type of glass  
count  196.000000  196.000000  196.000000     196.000000  
mean     8.910714    0.164235    0.050255       2.668367  
std      1.421490    0.485198    0.086359       2.062416  
min      5.430000    0.000000    0.000000       1.000000  
25%      8.220000    0.000000    0.000000       1.000000  
50%      8.575000    0.000000    0.000000       2.000000  
75%      9.092500    0.000000    0.090000       3.000000  
max     16.190000    3.150000    0.340000       7.000000

In [11]:
test_df.describe()

RI         Na         Mg         Al         Si          K  \
count  18.000000  18.000000  18.000000  18.000000  18.000000  18.000000   
mean    1.519131  13.763333   1.873889   1.342222  72.820000   0.253889   
std     0.002791   1.083491   1.749753   0.581312   0.892004   0.265133   
min     1.511150  12.850000   0.000000   0.340000  71.360000   0.000000   
25%     1.517412  13.012500   0.000000   0.877500  72.212500   0.000000   
50%     1.519515  13.355000   2.395000   1.320000  72.685000   0.175000   
75%     1.521055  14.220000   3.570000   1.902500  73.382500   0.502500   
max     1.523000  17.380000   3.780000   2.170000  75.410000   0.760000   

              Ca         Ba         Fe  Type of glass  
count  18.000000  18.000000  18.000000      18.000000  
mean    9.460556   0.292778   0.130556       4.000000  
std     1.380432   0.617422   0.164798       2.222876  
min     6.650000   0.000000   0.000000       1.000000  
25%     8.635000   0.000000   0.000000       2.000000  
50%     9.065000   0.000000   0.015000       4.000000  
75%    10.090000   0.112500   0.240000       6.000000  
max    12.500000   1.670000   0.510000       7.000000

Finding unique values for train and test data

In [12]:
train_df['Type of glass'].unique()

array([1, 2, 3, 5, 6, 7])

In [13]:
test_df['Type of glass'].unique()

array([1, 2, 3, 5, 6, 7])

In [14]:
from sklearn.preprocessing import StandardScaler

**Splitting the test and the train Data**

In [16]:
x_train=train_df.iloc[:,1:-1]
y_train=train_df['Type of glass']
x_test=test_df.iloc[:,1:-1]
y_test=test_df['Type of glass']

**Feature Scaling**

In [17]:
st=StandardScaler()
x_train=st.fit_transform(x_train)

In [18]:
x_test=st.fit_transform(x_test)

In [19]:
x_train

array([[ 0.3389847 ,  1.24616006, -0.72249922, ..., -0.11334995,
        -0.33935667, -0.5834224 ],
       [ 0.65902807,  0.60544985, -0.19235437, ..., -0.76221546,
        -0.33935667, -0.5834224 ],
       [ 0.19816562,  0.56945489,  0.174669  , ..., -0.79747989,
        -0.33935667, -0.5834224 ],
       ...,
       [ 1.27351132, -1.98618697,  2.62149143, ...,  0.38035207,
         0.77644048, -0.5834224 ],
       [ 0.97907143, -1.98618697,  2.90695404, ...,  0.18992414,
         1.8509118 , -0.5834224 ],
       [ 1.97760672, -1.98618697,  1.09222741, ..., -0.36020096,
         2.94604604, -0.5834224 ]])

In [20]:
x_test

array([[-0.67745268,  1.04449191, -0.83588893, -0.57678658, -0.24795484,
         0.29029673, -0.48794184,  0.24628886],
       [-0.61097368,  1.00332645, -0.7827854 , -0.71521535, -0.09271355,
         0.26793435, -0.48794184,  0.18384943],
       [-0.43053068,  1.00332645, -0.92439482, -0.95746572, -0.5196271 ,
         0.5288287 , -0.48794184, -0.62786315],
       [-0.72493768,  0.93863786,  0.79261939, -0.11535732,  1.57613032,
        -0.95454202, -0.48794184, -0.81518144],
       [-0.73443468,  0.76809524, -0.19864655,  0.17303597,  1.26564774,
        -0.48493219, -0.48794184,  0.68336487],
       [-0.86739269,  1.05625347, -0.18094537, -0.28839329,  1.42088903,
        -0.58183581, -0.48794184,  1.3701986 ],
       [-0.41153668,  0.86218772,  0.35008995, -0.78442974,  1.1880271 ,
        -0.35075796, -0.48794184, -0.81518144],
       [-0.11712967,  0.97980332, -1.04830306, -0.19610744, -0.36438581,
        -0.4252992 , -0.23795266,  0.68336487],
       [ 0.40520534,  1.12094205

#Euclidean Distance

In [21]:
euc_knn=KNeighborsClassifier(metric=distance.sqeuclidean)

In [22]:
#model created for Euclidean distance

euc_knn.fit(x_train,y_train)

KNeighborsClassifier(metric=<function sqeuclidean at 0x7f2a1fef8ca0>)

In [23]:
y_prediction_euc=euc_knn.predict(x_test)

In [24]:
y_prediction_euc

array([1, 1, 1, 2, 1, 2, 2, 2, 2, 5, 2, 5, 6, 2, 1, 7, 7, 7])

In [28]:
from sklearn.metrics import accuracy_score

In [29]:
#Finding accuracy level of the Euclidean prediction

accuracy_euc=accuracy_score(y_test,y_prediction_euc)*100
accuracy_euc

61.111111111111114

#Manhattan distance

In [35]:
#model created for Manhattan distance

manhattan_knn=KNeighborsClassifier(metric='manhattan')

In [36]:
manhattan_knn.fit(x_train,y_train)

KNeighborsClassifier(metric='manhattan')

In [37]:
y_prediction_manhattan=manhattan_knn.predict(x_test)

In [38]:
#Finding accuracy level of the Manhattan prediction

accuracy_manhattan=accuracy_score(y_test,y_prediction_manhattan)*100
accuracy_manhattan

66.66666666666666

In [40]:
#Ignoring warnings 

import warnings
warnings.filterwarnings('ignore')

**Classification Report for Euclidean and Manhattan**

In [43]:
print(classification_report(y_test,y_prediction_euc))
print("-----------------------------------------------------------------------")
print("The accuracy score of Euclidean Distance for the test Data :", accuracy_euc  )

              precision    recall  f1-score   support

           1       0.60      1.00      0.75         3
           2       0.29      0.67      0.40         3
           3       0.00      0.00      0.00         3
           5       1.00      0.67      0.80         3
           6       1.00      0.33      0.50         3
           7       1.00      1.00      1.00         3

    accuracy                           0.61        18
   macro avg       0.65      0.61      0.58        18
weighted avg       0.65      0.61      0.57        18

----------------------------------------
The accuracy score of Euclidean Distance for the test Data : 61.111111111111114


In [44]:
print(classification_report(y_test,y_prediction_manhattan))
print("-----------------------------------------------------------------------")
print("The accuracy score of Manhattan Distance for the test Data :", accuracy_manhattan)

              precision    recall  f1-score   support

           1       0.75      1.00      0.86         3
           2       0.29      0.67      0.40         3
           3       0.00      0.00      0.00         3
           5       1.00      0.67      0.80         3
           6       1.00      0.67      0.80         3
           7       1.00      1.00      1.00         3

    accuracy                           0.67        18
   macro avg       0.67      0.67      0.64        18
weighted avg       0.67      0.67      0.64        18

-----------------------------------------------------------------------
The accuracy score of Manhattan Distance for the test Data : 66.66666666666666


#CONCLUSION

The accuracy score of Manhattan Distance is 66.66 and the accuracy for Euclidean score is 61.11. This shows the model created for Manhattan Distance fits more in the given data